expression/expression.go

```diff
@@ -95,7 +113,49 @@ func (triggerExpression *TriggerExpression) Evaluate() (moira.State, error) {
 	}
 }
 
-func validateUserExpression(triggerExpression *TriggerExpression, userExpression *govaluate.EvaluableExpression) (*govaluate.EvaluableExpression, error) {
+// Validate gets trigger expression and validates it for given parameters using expr
+func (triggerExpression *TriggerExpression) Validate() error {
+	expression, err := getExpression(triggerExpression)
+	if err != nil {
+		return ErrInvalidExpression{internalError: err}
+	}
+
+	cacheKey := fmt.Sprintf("[VALIDATED]%s", expression.String())
+	pr, found := exprCache.Get(cacheKey)
+	program, ok := pr.(*vm.Program)
+	if !ok {
+		found = false
+		exprCache.Delete(cacheKey)
+	}
+	if !found {
+		program, err = expr.Compile(
+			expression.String(),
+			expr.Patch(triggerExpression), // patch identifiers with call to Get
+			expr.Optimize(true),
+		)
+		if err != nil {
+			return err
+		}
+		exprCache.Set(cacheKey, program, cache.NoExpiration)
+	}
+	result, err := expr.Run(program, map[string]interface{}{
```

Comment:
Should we do a compilation here? Let the faster `govaluate` do the execution of the expression

Reply:
`Run` needs to be called for `expr` to check further for validation. I'll give an example:
the compile stage changes `t1 => Get("t1")` but it's during the execution that the error returned by Get("t1") will be returned if "t1" does not exist.

The compilation stage only checks for basic syntax errors such as misplacing ternary operators.

Reply:
If `expr.Run()` checks absolutely all arguments, no problem, you can leave it like this